File: core/src/fundamentals/trait.rs
use crate::*;

#[derive(TypeInfo, Clone)]
pub enum Trait {
    Primitive(DynamicType),
    Runtime(Id, Pattern),
}

impl Trait {
    pub fn new(pattern: Pattern) -> Self {
        Trait::Runtime(Id::new(), pattern)
    }

    pub fn of<T: TypeInfo>() -> Self {
        Trait::Primitive(DynamicType::of::<T>())
    }
}

impl Trait {
    pub fn id(&self) -> Id {
        match self {
            Trait::Primitive(r#type) => Id::Primitive(*r#type),
            Trait::Runtime(id, _) => *id,
        }
    }

    pub fn pattern(&self) -> Pattern {
        match self {
            Trait::Primitive(_) => Pattern::for_trait(self.clone()),
            Trait::Runtime(_, pattern) => pattern.clone(),
        }
    }
}

impl PartialEq for Trait {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Trait::Primitive(a), Trait::Primitive(b)) => a == b,
            (Trait::Runtime(a, _), Trait::Runtime(b, _)) => a == b,
            _ => false,
        }
    }
}

impl Eq for Trait {}

impl std::fmt::Debug for Trait {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "Trait({:?})", self.id())
    }
}

impl Value {
    pub fn get_trait(&self, r#trait: &Trait, env: &Environment, stack: &Stack) -> Result {
        self.get_trait_or(r#trait, "Value does not have this trait", env, stack)
    }

    pub fn is_trait(&self, r#trait: &Trait, env: &Environment, stack: &Stack) -> Result<bool> {
        self.get_trait_if_present(r#trait, env, stack)
            .map(|t| t.is_some())
    }

    pub fn get_trait_or(
        &self,
        r#trait: &Trait,
        message: &str,
        env: &Environment,
        stack: &Stack,
    ) -> Result {
        self.get_trait_if_present(r#trait, env, stack)?
            .ok_or_else(|| Return::error(message, stack))
    }

    pub fn get_trait_if_present(
        &self,
        r#trait: &Trait,
        env: &Environment,
        stack: &Stack,
    ) -> Result<Option<Value>> {
        if let Some(value) = self.direct_value_for_trait(r#trait) {
            return Ok(Some(value.clone()));
        }

        self.derive(r#trait, env, stack)
    }
}

impl Value {
    pub fn get<T: TypeInfo>(&self, env: &Environment, stack: &Stack) -> Result<T> {
        self.get_or("Cannot find trait", env, stack)
    }

    pub fn get_or<T: TypeInfo>(
        &self,
        message: &str,
        env: &Environment,
        stack: &Stack,
    ) -> Result<T> {
        self.get_if_present(env, stack)?
            .ok_or_else(|| Return::error(message, stack))
    }

    pub fn get_if_present<T: TypeInfo>(
        &self,
        env: &Environment,
        stack: &Stack,
    ) -> Result<Option<T>> {
        Ok(self
            .get_trait_if_present(&Trait::of::<T>(), env, stack)?
            .map(|value| value.into_primitive::<T>().unwrap()))
    }
}

File: Language/Sources/Trait.swift
public enum Trait {
    case primitive(Any.Type)
    case runtime(ID, Pattern)

    public init(pattern: Pattern) {
        self = .runtime(ID(), pattern)
    }

    public init(_ type: Any.Type) {
        self = .primitive(type)
    }
}

public extension Trait {
    var id: ID {
        switch self {
        case .primitive(let type):
            return ID(type)
        case .runtime(let id, _):
            return id
        }
    }

    var pattern: Pattern {
        switch self {
        case .primitive:
            return Pattern(trait: self)
        case .runtime(_, let pattern):
            return pattern
        }
    }
}

extension Trait: Equatable {
    public static func == (lhs: Self, rhs: Self) -> Bool {
        switch (lhs, rhs) {
        case (.primitive(let lhs), .primitive(let rhs)):
            return lhs == rhs
        case (.runtime(let lhs, _), .runtime(let rhs, _)):
            return lhs == rhs
        default:
            return false
        }
    }
}

public extension Value {
    func `is`(_ trait: Trait, _ env: Env, _ stack: Stack) throws -> Bool {
        try self.getIfPresent(trait: trait, env, stack) != nil
    }

    func get(
        trait: Trait,
        or message: @autoclosure () -> String = "Value does not have this trait",
        _ env: Env,
        _ stack: Stack
    ) throws -> Value {
        guard let value = try self.getIfPresent(trait: trait, env, stack) else {
            throw Exit.error(message(), stack)
        }

        return value
    }

    func getIfPresent(trait: Trait, _ env: Env, _ stack: Stack) throws -> Value? {
        try self.directValue(for: trait) ?? self.derive(trait, env, stack)
    }
}

public extension Value {
    func `is`<T: Primitive>(_ type: T.Type, _ env: Env, _ stack: Stack) throws -> Bool {
        try self.is(Trait(type), env, stack)
    }

    func get<T: Primitive>(
        _ type: T.Type,
        or message: @autoclosure () -> String = "Value does not have this trait",
        _ env: Env,
        _ stack: Stack
    ) throws -> T {
        guard let value = try self.getIfPresent(type, env, stack) else {
            throw Exit.error(message(), stack)
        }

        return value
    }

    func getIfPresent<T: Primitive>(
        _ type: T.Type,
        _ env: Env,
        _ stack: Stack
    ) throws -> T? {
        try self.getIfPresent(trait: Trait(type), env, stack)?.primitiveValue as! T?
    }
}
